hw/ip/otbn/dv/model/otbn_model.cc

```diff
@@ -177,33 +181,106 @@ static void write_vector_to_file(const std::string &path,
   }
 }
 
-extern "C" OtbnModel *otbn_model_init(const char *mem_scope,
-                                      const char *design_scope,
-                                      unsigned imem_words,
-                                      unsigned dmem_words) {
-  assert(mem_scope && design_scope);
-  return new OtbnModel(mem_scope, design_scope, imem_words, dmem_words);
+// Extract 256-bit RND EDN data from a 4 state logic value. RND data is placed
+// into 8 element uint32_t array dst.
+static void set_rnd_data(uint32_t dst[8], const svLogicVecVal src[8]) {
+  for (int i = 0; i < 8; ++i) {
+    // All bits should be known
+    assert(src[i].bval == 0);
+    dst[i] = src[i].aval;
+  }
 }
 
-extern "C" void otbn_model_destroy(OtbnModel *model) { delete model; }
+static bool is_xz(svLogic l) { return l == sv_x || l == sv_z; }
+
+template <typename T>
+static std::array<T, 32> get_rtl_regs(const std::string &reg_scope) {
+  std::array<T, 32> ret;
+  static_assert(sizeof(T) <= 256 / 8, "Can only copy 256 bits");
+
+  SVScoped scoped(reg_scope);
+
+  // otbn_rf_peek passes data as a packed array of svBitVecVal words (for a
+  // "bit [255:0]" argument). Allocate 256 bits (= 32 bytes) as
+  // 32/sizeof(svBitVecVal) words on the stack.
+  svBitVecVal buf[256 / 8 / sizeof(svBitVecVal)];
+
+  // The implementation of otbn_rf_peek() is only available if DesignScope != ""
+  // (the function is implemented in SystemVerilog, and imported through DPI).
+  // We should not reach the code here if that's the case.
+  assert(otbn_rf_peek);
+
+  for (int i = 0; i < 32; ++i) {
+    if (!otbn_rf_peek(i, buf)) {
+      std::ostringstream oss;
+      oss << "Failed to peek into RTL to get value of register " << i
+          << " at scope `" << reg_scope << "'.";
+      throw std::runtime_error(oss.str());
+    }
+    memcpy(&ret[i], buf, sizeof(T));
+  }
+
+  return ret;
+}
+
+template <typename T>
+static std::vector<T> get_stack(const std::string &stack_scope) {
+  std::vector<T> ret;
+  static_assert(sizeof(T) <= 256 / 8, "Can only copy 256 bits");
+
+  SVScoped scoped(stack_scope);
+
+  // otbn_stack_element_peek passes data as a packed array of svBitVecVal words
+  // (for a "bit [255:0]" argument). Allocate 256 bits (= 32 bytes) as
+  // 32/sizeof(svBitVecVal) words on the stack.
+  svBitVecVal buf[256 / 8 / sizeof(svBitVecVal)];
+
+  // The implementation of otbn_stack_element_peek() is only available if
+  // DesignScope != "" (the function is implemented in SystemVerilog, and
+  // imported through DPI).  We should not reach the code here if that's the
+  // case.
+  assert(otbn_stack_element_peek);
+
+  int i = 0;
+
+  while (1) {
+    int peek_result = otbn_stack_element_peek(i, buf);
+
+    if (peek_result == 2) {
```

Comment:
Could you declare an enum for the return values for extra clarity?

Reply:
I could, although this crosses the DPI boundary, so then there's a chance for things to come adrift (which I think would probably be more confusing to debug than just raw numbers). Maybe it would be better just to add a comment above the call pointing at the (SV) implementation and explaining what the possible return values mean?

I've just done that, and slightly rejigged the code to add a missing range assertion. What do you think?

Reply:
Better, thanks.